### ISSUE TITLE
TFLite NCHW data format issue and Open Questions

### ISSUE THREAD

**Issue by pranathibl** [tensorflow/tensorflow#49381]
TFLite Version used - 2.4.1
Does TFlite Models can be created using NCHW format ?

When I tried to create , I was having issues ?
But the error i was getting was incorrect, Is this expected ?

In documentation it is not mentioned that NCHW is not supported ? When is this expected to support ?

Few open questions i found regarding this :- 
https://stackoverflow.com/questions/66671358/which-format-is-preferable-for-tflite-model-nchw-or-nhwc

Is it possible to create tflite nchw models ? When TFLite NCHW format is supported ? Facing issues in the attached TFLite script 

Open Questions :- 

**1.  Does TFLite supports NCHW ? 
2.  What is the default format of TFLite Models ? It is always N,H,W,C ?
3.  What is the default format of TFLite Models LSTM models ? It is always N,S,D ? Does TFLite LSTM support S,N,D ?
4.  What are the input formats other than  N,H,W, C  (or) N,S,D ?
5.  How is ideally LSTM's and CNN's differentiated in TFLite ?
6.  When will be NCHW supported for TFLite if its to be supported ?**

Facing issues in the below code :- 

```
import numpy as np
import tensorflow as tf
input_shape = (1, 3, 3, 227)
x = tf.random.normal(input_shape)


model = tf.keras.Sequential([
  tf.keras.layers.InputLayer(input_shape=( 1, 3,227,227)),
  tf.keras.layers.Conv2D(96, 11, activation='relu', strides=(4,4), dilation_rate=(1,1), groups=1, data_format = 'channels_first')])
model.save('my_conv_model') 

converter = tf.lite.TFLiteConverter.from_keras_model(model)

tflite_model = converter.convert()

with open('convmodel.tflite', 'wb') as f:
f.write(tflite_model)
```

**Comment by MeghnaNatraj** [comment]
I am unable to generate the TF model. Is it possible to fix this code? For questions about TensorFlow post a [TensorFlow bug](https://github.com/tensorflow/tensorflow/issues/new?assignees=&labels=type%3Abug&template=00-bug-issue.md) issue instead. We can then help you debug the conversion part.

**Comment by pranathibl** [comment]
Could you provide the solution to resolve this issue .
Please provide me the answers for the questions i asked.

**Comment by MeghnaNatraj** [comment]
1. Does TFLite supports NCHW ?
> No.
2. What is the default format of TFLite Models ? It is always N,H,W,C ?
> NHWC. Yes.
3. What is the default format of TFLite Models LSTM models ? It is always N,S,D ? Does TFLite LSTM support S,N,D ?
Could you point to an example that trains TF models with either of these formats?
4. What are the input formats other than N,H,W, C (or) N,S,D ?
Only NHWC is supported.
5. How is ideally LSTM's and CNN's differentiated in TFLite ?
It depends on the operator or layer type specified in the TF model.
6. When will be NCHW supported for TFLite if its to be supported ?
> This is not in our roadmap yet as there is no requirement.

**Comment by MeghnaNatraj** [comment]
I cannot provide a solution until I can first create a TensorFlow model. If you can provide a Tensorflow model along with the code, I can help with the conversion.

**Comment by pranathibl** [comment]
Yes i will point an example of LSTM which takes N(Batch Size), S(Sequence length) and D(features ) as input. As in previous reply, You are sayiong only N,H,W, C are supported , But for LSTM's it takes N,S,D. Is there any API to tell what kind of format input is ?

**Comment by pranathibl** [comment]
The below is the example of TFLite LSTM model 

`import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential()

model.add(layers.LSTM(input_shape=(None,20),units= 128, return_sequences = True))
model.add(layers.Dense(10))
model.summary()

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy']);
x_test = np.zeros([3000,4,20], dtype=np.float32);
x_train = np.zeros([3000,4,20], dtype=np.float32);
y_test =  np.zeros([3000, 4,], dtype=np.float32);
y_train = np.zeros([3000, 4,], dtype=np.float32);

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test, y_test)

run_model = tf.function(lambda x: model(x))

BATCH_SIZE = 1
STEPS = 2
INPUT_SIZE = 20

concrete_func = run_model.get_concrete_function(
    tf.TensorSpec([BATCH_SIZE, STEPS, INPUT_SIZE], model.inputs[0].dtype))


MODEL_DIR = "keras_lstm_new"
model.save(MODEL_DIR, save_format="tf", signatures=concrete_func)

converter = tf.lite.TFLiteConverter.from_saved_model(MODEL_DIR)
tflite_model = converter.convert()

with open('lstmmodelnew.tflite', 'wb') as f:
  f.write(tflite_model) 

`

**Comment by MeghnaNatraj** [comment]
From the documentation [here](https://www.tensorflow.org/lite/convert/rnn) and [example](https://colab.sandbox.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/experimental_new_converter/Keras_LSTM_fusion_Codelab.ipynb) NSD is supported. You can view the model and you'll notice that NSD is used. As the [API](https://github.com/tensorflow/tensorflow/blob/35a3ab91b42503776f428bda574b74b9a99cd110/tensorflow/python/keras/layers/recurrent_v2.py#L1238) also seems to use NSD  (see the Args section.. all tensors use [batch_size, time, units]), it looks like only NSD is currently supported.

NHWC was for CNN/image models.

**Comment by pranathibl** [comment]
Created issue in stack overflow - https://stackoverflow.com/questions/67857868/unable-to-create-tflite-model-with-time-major-format

Not sure, If we can create TFLite Models in [S,N,D] (time_major) format.

**Comment by pranathibl** [comment]
Any update ?

**Comment by MeghnaNatraj** [comment]
@renjie-liu

**Comment by renjie-liu** [comment]
Yes, you can create time-majored lstm model.
